src/class/cdc/cdc_device.c

```diff
@@ -432,25 +432,25 @@ bool cdcd_xfer_cb(uint8_t rhport, uint8_t ep_addr, xfer_result_t result, uint32_
   // Received new data
   if ( ep_addr == p_cdc->ep_out )
   {
+    tu_fifo_write_n(&p_cdc->rx_ff, &p_cdc->epout_buf, xferred_bytes);
+    
     // TODO search for wanted char first for better performance
-    for(uint32_t i=0; i<xferred_bytes; i++)
-    {
-      tu_fifo_write(&p_cdc->rx_ff, &p_cdc->epout_buf[i]);
-
+    if (tud_cdc_rx_wanted_cb && ( ((signed char) p_cdc->wanted_char) != -1)) {
       // Check for wanted char and invoke callback if needed
-      if ( tud_cdc_rx_wanted_cb && ( ((signed char) p_cdc->wanted_char) != -1 ) && ( p_cdc->wanted_char == p_cdc->epout_buf[i] ) )
-      {
-        tud_cdc_rx_wanted_cb(itf, p_cdc->wanted_char);
+      for (uint32_t i=0; i<xferred_bytes; i++) {
+        if ( p_cdc->wanted_char == p_cdc->epout_buf[i] ) {
+          tud_cdc_rx_wanted_cb(itf, p_cdc->wanted_char);
```

Comment:
hmm, maybe it is not worth to have `wanted` as the absolute last character. Maybe we can implement it this way. Though we may need to check if data is still available before invoking callback (much like the cdc_rx_cb() below) since previous callback can potentially read() all the data.

Reply:
After some thought I feel more hesitate about how to implement it :>

While I use UART extensively in our product, I always use a packet so there is no `last char` problem.

We also need to make sure data is written (fifo not full) before call the cb. If the application relies on `tud_cdc_rx_wanted_cb`, there could be a deadlock where the fifo is full but no wanted char received.

If it's a personal project I'd delete this cb and let the application to do all the decoding :) I think we can ask more people about what they think the best way to do it.

Reply:
Oh, I didn't think about the edge case where fifo is full and wanted char has yet to come. Maybe we should have an `rx_full_cb()`. However, it should not be affected by existing code or your PR. After all, if application want to process a large data with wanted ch (e.g command with newline), they should tune up the fifo to large enough. Anyway, I think we should just continue with this PR, would you mind to add the check if fifo not empty before invoked the wanted callback. That would be enough for the merge. If needed the rx_full_cb() can be added as separated PR.

Reply:
I've added fifo empty check in last commit.